libs/shared/components/src/Inputs/TokenInput.tsx

```diff
@@ -0,0 +1,178 @@
+import { forwardRef } from 'react';
+
+import { alpha, Box, Button, Skeleton, Stack, Typography } from '@mui/material';
+import { useIntl } from 'react-intl';
+import { formatUnits } from 'viem';
+
+import { BigintInput } from './BigintInput';
+
+import type { StackProps } from '@mui/material';
+import type { Token } from '@origin/shared/contracts';
+
+import type { BigintInputProps } from './BigintInput';
+
+export type TokenInputProps = {
+  amount: bigint;
+  decimals?: number;
+  onAmountChange?: (value: bigint) => void;
+  isAmountLoading?: boolean;
+  isAmountDisabled?: boolean;
+  isAmountError?: boolean;
+  isConnected: boolean;
+  balance?: bigint;
+  isBalanceLoading?: boolean;
+  disableMaxClick?: boolean;
+  token: Token;
+  onTokenClick: () => void;
+  tokenPriceUsd?: number;
+  isPriceLoading?: boolean;
+  inputProps?: Omit<
+    BigintInputProps,
+    'value' | 'decimals' | 'onChange' | 'isLoading' | 'isError'
+  >;
+} & StackProps;
+
+export const TokenInput = forwardRef<HTMLInputElement, TokenInputProps>(
+  (
+    {
+      amount,
+      decimals = 18,
+      onAmountChange,
+      isAmountLoading,
+      isAmountDisabled,
+      isAmountError,
+      isConnected,
+      balance = 0n,
+      isBalanceLoading,
+      disableMaxClick,
+      token,
+      onTokenClick,
+      tokenPriceUsd = 0,
+      isPriceLoading,
+      inputProps,
+      ...rest
+    },
+    ref,
+  ) => {
+    const intl = useIntl();
+
+    const handleMaxClick = () => {
+      if (onAmountChange) {
+        onAmountChange(balance);
+      }
+    };
+
+    const bal = +formatUnits(balance, decimals);
+    const amountUsd = +formatUnits(amount, decimals) * tokenPriceUsd;
+
+    return (
+      <Stack direction="row" {...rest}>
+        <Stack flexGrow={1} justifyContent="center" gap={1}>
+          <BigintInput
+            {...inputProps}
+            value={amount}
+            decimals={decimals}
+            onChange={onAmountChange}
+            ref={ref}
+          />
+          {isPriceLoading ? (
+            <Skeleton width={30} />
+          ) : tokenPriceUsd > 0 ? (
+            <Typography variant="body2">
+              {intl.formatNumber(amountUsd, {
```

Comment:
We should probably expose some shared helpers for number formatters even if just proxying